src/main/java/core/basesyntax/ArrayList.java

```
@@ -1,48 +1,170 @@
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int s = size;
+            if (s == array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index,
+                array, index + 1,
+                s - index);
+            array[index] = value;
+            size = s + 1;
+        }
     }
 
+
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            T[] tempArray = (T[]) listToArray(list);
+            int newCapacity = size + tempArray.length;
+            array = Arrays.copyOf(array, newCapacity);
+            System.arraycopy(tempArray, 0, array, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
+
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            return array[index];
+        }
     }
 
+
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            array[index] = value;
+        }
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index);
+            }
+            size = newSize;
+            array[newSize] = null;
+            return oldValue;
+        }
     }
 
+
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found:
```

Using labels is a very bad practice, it's like a `goto` mechanism, which is not recommended by most the programming languages

---

I think the better idea might be to create a separate private method that returns the index of a found element. If it's null -> you can throw the exception, otherwise - do the remove by index to avoid the duplication of `remove(index);`